### PR TITLE
[WFLY-6821] Not possible to overlay non existing file in WAR - mark the sub deployment which will be mounted exploded

### PR DESCRIPTION
**Pull request by gaol**
Jira: https://issues.jboss.org/browse/WFLY-6821

The linked WildFly core Jira: https://issues.jboss.org/browse/WFCORE-761

This is second part of fix and the test cases on overlay issues, the test cases contain following situations:

1). replace/new files in jar, like: ejb.jar/META-INF/jboss-ejb3.xml
2). replace/new files in war, like: web.war/index.html
3). replace/new files in war/jar, like: web.war/WEB-INF/lib/mylib.jar/META-INF/config.properties
4). replace/new files in ear/jar, like: app.ear/lib/earlib.jar/META-INF/config.properties
5). replace/new files in ear/war, like: app.ear/web.war/index.html
~~6). replace/new files in ear/war/jar, like: app.ear/web.war/WEB-INF/lib/mylib.jar/META-INF/config.properties~~

Test case No.6 above is disabled now, because PR in wildfly-core does not fix this case, and it needs more discussion.

### CODE REVIEW

**Comment by gaol** [comment]
Please hold on the tests, it needs tests together with changes in wildfly-core

**Comment by gaol** [comment]
Enable test case number 6 in the description: 
6). replace/new files in ear/war/jar, like: app.ear/web.war/WEB-INF/lib/mylib.jar/META-INF/config.properties

PR: https://github.com/wildfly/wildfly-core/pull/1261 will override PR: wildfly/wildfly-core#1332, and overlay should not have any limitation anymore, this test case should pass too.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/ear/EarOverlayTestBase.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        @Override
+        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+            InputStream input = null;
+            try {
+                input = getClass().getClassLoader().getResourceAsStream(META_RESOURCE_IN_JAR_OVERLAY);
+                if (input == null) {
+                    resp.sendError(404);
+                    return;
+                }
+                BufferedReader reader = new BufferedReader(new InputStreamReader(input));
```
Comment (ctomc) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/ear/EarOverlayTestBase.java]: resource leak

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/ear/EarOverlayTestBase.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    @SuppressWarnings("serial")
+    @WebServlet("/echoStatic")
+    public static class EchoStaticServlet extends HttpServlet {
+
+        @Override
+        protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+            InputStream input = null;
+            try {
+                input = getClass().getClassLoader().getResourceAsStream(META_RESOURCE_IN_JAR_STATIC);
+                BufferedReader reader = new BufferedReader(new InputStreamReader(input));
```
Comment (ctomc) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/ear/EarOverlayTestBase.java]: resource leak

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/ear/OverlayNonExistingResourceTestCase.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+        Assert.assertEquals("HTML static resource in ear/war does not match post-overlay expectations!", OverlayableInterface.STATIC,
+                readContent(managementClient.getWebUri() + "/" + WEB + "/" + STATIC_HTML));
+        Assert.assertEquals("HTML resource in ear/war does not match post-overlay expectations!", OverlayableInterface.OVERLAYED,
+                readContent(managementClient.getWebUri() + "/" + WEB + "/" + OVERLAY_HTML));
+
+        //Assert.assertEquals("Static resource in ear/war/jar does not match post-overlay expectations!", OverlayableInterface.STATIC,
+        //        readContent(managementClient.getWebUri() + "/" + WEB + "/echoStatic"));
+        //Assert.assertEquals("Resource in ear/war/jar does not match post-overlay expectations!", OverlayableInterface.OVERLAYED,
+        //        readContent(managementClient.getWebUri() + "/" + WEB + "/echoOverlay"));
```
Comment (ctomc) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/ear/OverlayNonExistingResourceTestCase.java]: remove comments

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/jar/OverlayUtils.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                removeContentItem(managementClient, overlayName, overlayPath);
+            }
+            removeDeploymentItem(managementClient, overlayName, deployment);
+
+            ModelNode op = new ModelNode();
+            op.get(ModelDescriptionConstants.OP_ADDR).set(ModelDescriptionConstants.DEPLOYMENT_OVERLAY, overlayName);
+            op.get(ModelDescriptionConstants.OP).set(ModelDescriptionConstants.REMOVE);
+            ManagementOperations.executeOperation(managementClient.getControllerClient(), op);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (ctomc) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/jar/OverlayUtils.java]: exception is swallowed.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/war/WarOverlayTestBase.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    public static String readContent(String url) throws IOException {
+        HttpURLConnection conn = null;
+        BufferedReader reader = null;
+        try {
+            conn = (HttpURLConnection)new URL(url).openConnection();
+            conn.connect();
+            int httpCode = conn.getResponseCode();
+            if (httpCode == 200) {
+                reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
```
Comment (ctomc) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/war/WarOverlayTestBase.java]: leak

Comment (gaol) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/war/WarOverlayTestBase.java]: can you tell why there is leak in this place please?

**Comment by gaol** [comment]
Updated to fix the potential resource leak but the last one.  :-)

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/war/WarOverlayTestBase.java (hunk 5 above)
Comment (ctomc) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/war/WarOverlayTestBase.java]: you have 3 resources opened in this line, but you only close the most outer one (reader)
easiest fix would be to use try-with-resources
try (BufferedReader  reader = new BufferedReader(new InputStreamReader(conn.getInputStream())){
..

Comment (gaol) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/deployment/deploymentoverlay/war/WarOverlayTestBase.java]: Will the `close()` method in most outer reader close the inner resources?

**Comment by gaol** [comment]
@ctomc, I see the problem.  It depends on the implementation of each level of stream/reader that the `close()` method will close nested streams too. Thanks.

I updated PR using `try-with-resource` with all opened resources. Could you please check again? Thank you.

**Comment by gaol** [comment]
@ctomc, I see the problem.  It depends on the implementation of each level of stream/reader that the `close()` method will close nested streams too. Thanks.

I updated PR using `try-with-resource` with all opened resources. Could you please check again? Thank you.

**Comment by ctomc** [comment]
@gaol looks better, tnx!

**Comment by kabir** [comment]
Retest this please

**Comment by kabir** [comment]
Retest this please

**Comment by gaol** [comment]
@kabir Hi, this needs master branch of wildfly-core to compile,  [The commit](https://github.com/wildfly/wildfly-core/pull/1332/commits/253faebd8ba1d629f1fc55edb44c0d26f4804622) has been merged in version `3.0.0.Alpha5`, but not in version: `3.0.0.Alpha4`.

**Comment by bstansberry** [comment]
retest this please

**Pull request merged**